File: 6.栈与队列/6.逆波兰表达式求值/150.逆波兰表达式求值.cpp
/*
 * @lc app=leetcode.cn id=150 lang=cpp
 *
 * [150] 逆波兰表达式求值
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> stk;
        int res = 0;
        for (auto s : tokens)
        {
            if (s != "+" && s != "-" && s != "*" && s != "/")
            {
                stk.push(stoi(s));
            }
            else
            {
                int a = stk.top();
                stk.pop();
                int b = stk.top();
                stk.pop();
                int tmp = 0;
                if (s == "+") tmp = b + a;
                else if (s == "-") tmp = b - a;
                else if (s == "*") tmp = b * a;
                else tmp = b / a;
                stk.push(tmp);
            }
        }
        return stk.top();
    }
};
// @lc code=end

File: 面经150/面经150p1/7.栈/150.逆波兰表达式求值.py
#
# @lc app=leetcode.cn id=150 lang=python3
#
# [150] 逆波兰表达式求值
#
from typing import List
# @lc code=start
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stk = []
        for t in tokens:
            if t.lstrip("+-").isdigit(): stk.append(int(t))
            else:
                a, b = stk.pop(), stk.pop()
                if t == '+': stk.append(b + a)
                elif t == '-': stk.append(b - a)
                elif t == '*': stk.append(b * a)
                else: stk.append(int(b / a))
        return stk[-1]
# @lc code=end

